qualtran/bloqs/rotations/phasing_via_cost_function.py

```diff
@@ -0,0 +1,136 @@
+#  Copyright 2023 Google LLC
+#
+#  Licensed under the Apache License, Version 2.0 (the "License");
+#  you may not use this file except in compliance with the License.
+#  You may obtain a copy of the License at
+#
+#      https://www.apache.org/licenses/LICENSE-2.0
+#
+#  Unless required by applicable law or agreed to in writing, software
+#  distributed under the License is distributed on an "AS IS" BASIS,
+#  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+#  See the License for the specific language governing permissions and
+#  limitations under the License.
+from functools import cached_property
+from typing import Dict, TYPE_CHECKING, Union
+
+import attrs
+import numpy as np
+import sympy
+
+from qualtran import Bloq, BloqBuilder, GateWithRegisters, QFxp, Register, Signature
+from qualtran.bloqs.basic_gates.rotation import ZPowGate
+from qualtran.bloqs.rotations.phase_gradient import AddScaledValIntoPhaseReg
+
+if TYPE_CHECKING:
+    from qualtran import SoquetT
+
+
+@attrs.frozen
+class PhasingViaCostFunction(Bloq):
```

Comment:
docstrings are missing for these classes.

Reply:
Also I'm not sure how strictly we want to enforce it, but the autodoc / test feature will have to be added at some point.

Reply:
Added detailed docstrings. PTAL